### PR TITLE
Update Spring versions

### PR DESCRIPTION
**Pull request by bogdan-sava**

### CODE REVIEW

**Comment by planetf1** [comment]
I think we should exchange notes on the different proposals before merging. I've tried to summarise what I had tried in the referenced PR above.

**Comment by planetf1** [comment]
I tried this out on macOS

* We are using spring boot 2.1.9 here - not 2.2.0
* The UI fails to start (protocol handler start failed) [this is after a reboot/nothing running on port 8443)
* We have a mix of versions - 2.1.3 & 2.1.11 both surface for ipa, and we have a series of additional properties added for different subcomponents - for example holding webmvc back at 5.1.10 but leaving others at 5.2.0

In summary I think we've both taken slightly different routes/targets ... but there's good ideas in both. Next week perhaps we can figure out how to connect the two strands together & ensure we have worked through some testing (including cohorts, containers etc) as this is a very fundamental change.

I'll continue to add notes in either PR/issue as I find anything new out.

**Comment by bogdan-sava** [comment]
This one is tested for both server chassis and UI

**Comment by bogdan-sava** [comment]
spring boot 2.2.0 doesn't work with webmvc 5.2.0 :)

**Comment by planetf1** [comment]
We noticed some problems here that we have been debugging - but also see issues. We appeared to find a potential working config in the related issue (though I've not proven this fixes webmvc yet - it compiles but need to test more... was blocked by 

I think we should hold off updates until we can get to the bottom of it and then recommend an approach - be that a partial update, our full solution, ignoring updates etc.

Could we move this PR to draft status (not sure if poss) as we work in finding out a good combination? It would prevent accidental inclusion?

**Comment by planetf1** [comment]
Unfortunately I don't think #1774 addresses all the updates either. I'll add some review comments there.

Ref webmvc -- do you have a specific reference to this being stated or is it based on experimentation? Both are valid, just trying to understand.

File: pom.xml [hunk 1]
```diff
@@ -131,18 +131,24 @@
...
-        <slf4j.version>1.7.26</slf4j.version>
+        <slf4j.version>1.7.28</slf4j.version>
         <logback.version>1.2.3</logback.version>
         <jackson.version>2.9.9</jackson.version>
         <jackson.databind.version>2.9.9.3</jackson.databind.version>
         <kafka.version>2.1.1</kafka.version>
         <janus.version>0.3.1</janus.version>
         <gremlin.version>3.3.3</gremlin.version>
-        <spring-boot.version>2.1.6.RELEASE</spring-boot.version>
+        <spring-boot.version>2.1.9.RELEASE</spring-boot.version>
```
Comment (planetf1) [inline pom.xml]: The latest is 2.2.0 - the challenge is if we can find suitable interdependent versions. (see discussion on mvc in the overall PR discussion)

File: pom.xml [hunk 2]
```diff
@@ -131,18 +131,24 @@
...
         <spring-data.version>2.1.9.RELEASE</spring-data.version>
-        <spring-web.version>5.1.8.RELEASE</spring-web.version>
-        <spring-security.version>5.1.5.RELEASE</spring-security.version>
-        <tomcat.version>9.0.21</tomcat.version>
+        <spring-jdbc.version>5.1.8.RELEASE</spring-jdbc.version>
```
Comment (planetf1) [inline pom.xml]: springjdbc follows the main spring versioning so we have to be very clear on why we are making an exception (presumably we use this setting of course). It also means we are not picking up other fixes.

File: pom.xml [hunk 3]
```diff
@@ -131,18 +131,24 @@
...
+        <spring-boot-jpa.version>2.1.3.RELEASE</spring-boot-jpa.version>
```
Comment (planetf1) [inline pom.xml]: The -jpa versions are aligned with spring-data & spring-boot respectively. We need to clarify exactly why we are going out of sync on these versions as there could be other subtle issues (IMO!) by doing so.

File: pom.xml [hunk 4]
```diff
@@ -131,18 +131,24 @@
...
+        <spring-data-jpa.version>2.1.11.RELEASE</spring-data-jpa.version>
+        <spring.version>5.2.0.RELEASE</spring.version>
+        <spring-web.version>5.2.0.RELEASE</spring-web.version>
+        <spring-webmvc.version>5.1.10.RELEASE</spring-webmvc.version>
```
Comment (planetf1) [inline pom.xml]: Same for spring mvc. Do we have a known issue, or documentation that forces us to keep back level. We need to clarify.

File: pom.xml [hunk 5]
```diff
@@ -131,18 +131,24 @@
...
+        <spring-expression.version>5.2.0.RELEASE</spring-expression.version>
+        <spring-security.version>5.2.0.RELEASE</spring-security.version>
+        <tomcat.version>9.0.26</tomcat.version>
```
Comment (planetf1) [inline pom.xml]: now at 9.0.27 - no doubt debendabot will point this out.

File: pom.xml [hunk 6]
```diff
@@ -131,18 +131,24 @@
...
-        <spring-web.version>5.1.8.RELEASE</spring-web.version>
-        <spring-security.version>5.1.5.RELEASE</spring-security.version>
-        <tomcat.version>9.0.21</tomcat.version>
+        <spring-jdbc.version>5.1.8.RELEASE</spring-jdbc.version>
+        <spring-boot-jpa.version>2.1.3.RELEASE</spring-boot-jpa.version>
+        <spring-data-jpa.version>2.1.11.RELEASE</spring-data-jpa.version>
+        <spring.version>5.2.0.RELEASE</spring.version>
+        <spring-web.version>5.2.0.RELEASE</spring-web.version>
+        <spring-webmvc.version>5.1.10.RELEASE</spring-webmvc.version>
+        <spring-expression.version>5.2.0.RELEASE</spring-expression.version>
```
Comment (planetf1) [inline pom.xml]: since expression is part of the core spring framework, why do we need an additional property - something else to update or get wrong.

File: pom.xml [hunk 7]
```diff
@@ -1767,112 +1773,115 @@
...
-                <scope>compile</scope>
                 <version>${spring-boot.version}</version>
             </dependency>
 
             <dependency>
                 <groupId>org.springframework.boot</groupId>
                 <artifactId>spring-boot-starter-data-jpa</artifactId>
-                <scope>compile</scope>
-                <version>${spring-boot.version}</version>
+                <version>${spring-boot-jpa.version}</version>
```
Comment (planetf1) [inline pom.xml]: see above comment on the properties

File: pom.xml [hunk 8]
```diff
@@ -1767,112 +1773,115 @@
...
                 <groupId>org.springframework.data</groupId>
                 <artifactId>spring-data-jpa</artifactId>
-                <scope>compile</scope>
                 <version>${spring-data.version}</version>
             </dependency>
 
+            <dependency>
+                <groupId>org.springframework</groupId>
+                <artifactId>spring-jdbc</artifactId>
+                <version>${spring-jdbc.version}</version>
```
Comment (planetf1) [inline pom.xml]: see above comments on why jdbc specific version

File: pom.xml [hunk 9]
```diff
@@ -1767,112 +1773,115 @@
...
                 <groupId>org.springframework</groupId>
                 <artifactId>spring-web</artifactId>
-                <scope>compile</scope>
                 <version>${spring-web.version}</version>
             </dependency>
 
+            <dependency>
+                <groupId>org.springframework</groupId>
+                <artifactId>spring-expression</artifactId>
+                <version>${spring-expression.version}</version>
```
Comment (planetf1) [inline pom.xml]: as above - unique version

File: pom.xml [hunk 10]
```diff
@@ -1767,112 +1773,115 @@
...
+            </dependency>
+
             <dependency>
                 <groupId>org.springframework</groupId>
                 <artifactId>spring-webmvc</artifactId>
-                <version>${spring-web.version}</version>
+                <version>${spring-webmvc.version}</version>
```
Comment (planetf1) [inline pom.xml]: as above - unique version

File: pom.xml [hunk 11]
```diff
@@ -131,18 +131,24 @@
...
+        <spring-boot.version>2.1.9.RELEASE</spring-boot.version>
         <spring-data.version>2.1.9.RELEASE</spring-data.version>
-        <spring-web.version>5.1.8.RELEASE</spring-web.version>
-        <spring-security.version>5.1.5.RELEASE</spring-security.version>
-        <tomcat.version>9.0.21</tomcat.version>
+        <spring-jdbc.version>5.1.8.RELEASE</spring-jdbc.version>
+        <spring-boot-jpa.version>2.1.3.RELEASE</spring-boot-jpa.version>
+        <spring-data-jpa.version>2.1.11.RELEASE</spring-data-jpa.version>
+        <spring.version>5.2.0.RELEASE</spring.version>
+        <spring-web.version>5.2.0.RELEASE</spring-web.version>
```
Comment (planetf1) [inline pom.xml]: we started using spring-web. I agree spring is a better name - but these are from the same release train - the core spring framework, so we should not use two properties. Fine to rename, but I disagree with having two without an explanation.

File: pom.xml (hunk 1 above)
Comment (bogdan-sava) [reply pom.xml]: done

File: pom.xml (hunk 3 above)
Comment (bogdan-sava) [reply pom.xml]: I agree.
Please check again

File: pom.xml (hunk 4 above)
Comment (bogdan-sava) [reply pom.xml]: I agree.
PLease check again

File: pom.xml (hunk 5 above)
Comment (bogdan-sava) [reply pom.xml]: done

File: pom.xml (hunk 7 above)
Comment (bogdan-sava) [reply pom.xml]: done

File: pom.xml (hunk 8 above)
Comment (bogdan-sava) [reply pom.xml]: done

File: pom.xml (hunk 9 above)
Comment (bogdan-sava) [reply pom.xml]: yup :)

**Comment by bogdan-sava** [comment]
@planetf1 please check again.
I have updated the versions and tested UI and server-chassis

**Comment by bogdan-sava** [comment]
spring offers dependencies that basically bundles dependencies needed for different kind of projects, and I looked at the dependency tree for them, build it and test it. As we are using managed dependencies we need to specify them.
Examples could be generated with start.spring.io

**Comment by bogdan-sava** [comment]
Now there are only 2 versions :
2.2.0 for spring-boot
5.2.0 for spring

**Comment by bogdan-sava** [comment]
Don't know how to move it to draft :)

**Comment by planetf1** [comment]
Thanks Bogdan - these latest changes look much cleaner. I will aim to test in the next few hours.

**Comment by planetf1** [comment]
I noticed a similar issue in my experimentation - I needed to update to the jakarta bundle for compile time, but this resulted in the runtime issue. We may need jab added at runtime also.

**Comment by planetf1** [comment]
I repeated the tests today
 - merged in with other pom changes
 - Added:

<dependency>
    <groupId>javax.xml.bind</groupId>
    <artifactId>jaxb-api</artifactId>
    <version>2.3.1</version>
</dependency>

[ note- this will need to go in dependency management etc following usual pattern ]

With this done I was able to run through configuring the servers and cohort with multiple members. I went through the asset catalog notebook. 

The UI also launched successfully, and I could run TEX. I did not try asset search.

Unfortunately glossary OMAS seemed to be failing - blank page - but I've seen this before on other changes, needs a little more investigation and may or may not be related.

Can I suggest you merge in latest master, add the java xml bind update as a runtime dependency (it will need an exclusion) and then we can both retest?

**Comment by planetf1** [comment]
Hi,
 I pulled your PR a few moments ago & retried.

I've captured the changes as a PR against your fork so that you can look at/review/include. It should just apply. I can also create a patch if it helps.

I have
 - merged in the correct current master pom changes to resolve conflicts
 - added javax.xml.bind

I have also confirmed
 - Chassis starts up ok
 - UI starts up ok
 - Can run through notebooks for configuration, server start, cohort membership, and asset catalog
 - can login to GUI
 - can run TEX

The only failure is that the subject area glossary page remains blank - but I see that regularly and wonder if it's work in progress by @davidradl 

Happy to submit a PR if easier - but see you may have just merged changes anyway, Hopefully we can get that in soon.

If you have a PR soon let me know and I can run tests again then we can merge

**Comment by planetf1** [comment]
I see the latest code changes are including some securityconfig updates - would this affect. https://github.com/odpi/egeria/issues/1812 (in a positive way?) cc: @ruxandraRosu

**Comment by planetf1** [comment]
Have tried the PR which just built now. I get:
```
Error starting ApplicationContext. To display the conditions report re-run your application with 'debug' enabled.
2019-10-24 13:20:44.655 ERROR 63343 --- [           main] o.s.boot.SpringApplication               : Application run failed

org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'entityManagerFactory' defined in class path resource [org/springframework/boot/autoconfigure/orm/jpa/HibernateJpaConfiguration.class]: Invocation of init method failed; nested exception is java.lang.NoClassDefFoundError: javax/xml/bind/JAXBException
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1803) ~[spring-beans-5.2.0.RELEASE.jar!/:5.2.0.RELEASE]

```
on starting the UI

Does it work for you?

**Comment by ruxandraRosu** [comment]
> #1812

Hi, yes, Bogdan pushed a fix for the issue.

**Comment by bogdan-sava** [comment]
@planetf1 you just get the build between the pushes :)

**Comment by planetf1** [comment]
I will test the latest PR - may be tomorrow now. Nearly there :-)

**Comment by planetf1** [comment]
Still fails to launch the UI for me I'm afraid
```
2019-10-24 19:12:41.807  INFO 66846 --- [           main] org.hibernate.Version                    : HHH000412: Hibernate Core {5.4.6.Final}
2019-10-24 19:12:42.049  WARN 66846 --- [           main] ConfigServletWebServerApplicationContext : Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'entityManagerFactory' defined in class path resource [org/springframework/boot/autoconfigure/orm/jpa/HibernateJpaConfiguration.class]: Invocation of init method failed; nested exception is java.lang.NoClassDefFoundError: javax/xml/bind/JAXBException
2019-10-24 19:12:42.050  INFO 66846 --- [           main] com.zaxxer.hikari.HikariDataSource       : HikariPool-1 - Shutdown initiated...
2019-10-24 19:12:42.054  INFO 66846 --- [           main] com.zaxxer.hikari.HikariDataSource       : HikariPool-1 - Shutdown completed.
2019-10-24 19:12:42.059  INFO 66846 --- [           main] o.apache.catalina.core.StandardService   : Stopping service [Tomcat]
2019-10-24 19:12:42.074  INFO 66846 --- [           main] ConditionEvaluationReportLoggingListener :

Error starting ApplicationContext. To display the conditions report re-run your application with 'debug' enabled.
2019-10-24 19:12:42.084 ERROR 66846 --- [           main] o.s.boot.SpringApplication               : Application run failed

org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'entityManagerFactory' defined in class path resource [org/springframework/boot/autoconfigure/orm/jpa/HibernateJpaConfiguration.class]: Invocation of init method failed; nested exception is java.lang.NoClassDefFoundError: javax/xml/bind/JAXBException
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1803) ~[spring-beans-5.2.0.RELEASE.jar!/:5.2.0.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:595) ~[spring-beans-5.2.0.RELEASE.jar!/:5.2.0.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:517) ~[spring-beans-5.2.0.RELEASE.jar!/:5.2.0.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:323) ~[spring-beans-5.2.0.RELEASE.jar!/:5.2.0.RELEASE]
```

It worked ok with the patch I made - but I did add jaxb for runtime.

**Comment by planetf1** [comment]
Does not appear to affect java 8.... I think it is due to the removal from the jdk of that function. We therefore need to add the library and test it works in both envs. The code must work in 8 and 11 now.

**Comment by planetf1** [comment]
see https://www.jesperdj.com/2018/09/30/jaxb-on-java-9-10-11-and-beyond/

**Comment by bogdan-sava** [comment]
I've installed 5 versions of openjdk now.
Yes! Indeed. This is weird.
I'll fix it.

Please note that as far as I can see openjdk 11 is no longer supported by openjdk. Why are we support it ?

**Comment by bogdan-sava** [comment]
@planetf1 It is fixed also for openjdk11

**Comment by planetf1** [comment]
Java 11 is the LTS release. The adoptopenjdk project has a good summary at https://adoptopenjdk.net/support.html . Oracle's page is at https://www.oracle.com/technetwork/java/java-se-support-roadmap.html

Java 12, 13 have come out, but it neither replace 11 as LTS yet - and it is these releases we should target (only). Of course it's good to check we can still build/run with 12,13 etc but it's 8,11 that are critical.

openjdk is distributed in a number of ways - by adoptopenjdk (download and via homebrew), by virtually every linux distribution, through docker images (including jdk), by the cloud providers (like Microsoft - we use their built version for our builds). It should only be packaging that differs pretty much !

I'll give the fix a go now. Thanks!

**Comment by planetf1** [comment]
Tests went well. I was able to run through build on 8 & 11, initialisation, UI (except SA - outstanding issue), server config, start (3 cluster notebook config), asset management (CSV files). 
great to get this through. Have now merged.

**Comment by planetf1** [comment]
One minor note. See #1774 notes referring to spring fox. We do still have spring-plugin-core & spring-plugin-metadata at 1.2 (instead of 2.0). Slightly aside from the core spring modules, but an update to 2.0 does break spring fox . Also option to move from spring fox - see https://github.com/odpi/egeria/issues/1124
